lib/hcrypto/hmac.c

```diff
@@ -80,49 +93,39 @@ HMAC_Init_ex(HMAC_CTX *ctx,
 	     ENGINE *engine)
 {
     unsigned char *p;
-    size_t i;
+    size_t i, blockSize;
+
+    blockSize = EVP_MD_block_size(md);
 
     if (ctx->md != md) {
+        if (ctx->md != NULL)
+            HMAC_CTX_cleanup(ctx);
+
 	ctx->md = md;
-	if (ctx->buf) {
-	    memset(ctx->buf, 0, ctx->key_length);
-	    free (ctx->buf);
-	}
 	ctx->key_length = EVP_MD_size(ctx->md);
 	ctx->buf = malloc(ctx->key_length);
+        ctx->opad = malloc(blockSize);
```

Comment:
Suppose you could check the return value of malloc but – it's no worse than the code that was already there!

Reply:
I've actually got another change (not pushed yet) that makes all of these buffers go away. The compilers I have tested generate much more efficient code when the size of buf, opad and ipad are known at compilation time, and it's worth the overhead of setting these to the maximum known hash block size and doing more iterations in order to get the more efficient code.